accepted/2023/wasm-browser-threads.md

```diff
@@ -0,0 +1,268 @@
+# Multi-threading on a browser
+
+## Goals
+ - CPU intensive workloads on dotnet thread pool
+ - enable blocking .Wait APIs from C# user code on all threads
+     - Current public API throws PNSE for it
+     - This is core part on MT value proposition.
+     - If people want to use existing MT code-bases, most of the time, the code is full of locks. 
+     - People want to use existing desktop/server multi-threaded code as is.
+ - allow HTTP and WS C# APIs to be used from any thread
+     - Underlying JS object have thread affinity
+ - JSImport/JSExport interop in maximum possible extent
+ - don't change/break single threaded build. †
+
+## Lower priority goals
+ - try to make it debugging friendly
+ - implement crypto via `subtle` browser API
+ - allow calls to synchronous JSExport from UI thread (callback)
+
+<sub><sup>† Note: all the text below discusses MT build only, unless explicit about ST build.</sup></sub>
+
+## Context - Problems
+**1)** If you have multithreading, any thread might need to block while waiting for any other to release a lock.
+     - locks are in the user code, in nuget packages, in Mono VM itself
+     - there are managed and un-managed locks
+     - in single-threaded build of the runtime, all of this is NOOP. That's why it works on UI thread.
+
+**2)** UI thread in the browser can't synchronously block
+    - you can spin-lock but it's bad idea.
+        - Deadlock: when you spin-block, the JS timer loop and any messages are not pumping. But code in other threads may be waiting for some such event to resolve.
+        - It eats your battery
+        - Browser will kill your tab at random point (Aw, snap).
+        - It's not deterministic and you can't really test your app to prove it harmless.
+    - Firefox (still) has synchronous XHR which could be captured by async code in service worker
```

Comment:
What is XHR?

Reply:
>What is XHR?


https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest

Reply:
Updated the section, 
https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Synchronous_and_Asynchronous_Requests#synchronous_request